velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h

```diff
@@ -0,0 +1,487 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <algorithm>
+#include <string>
+
+#include "velox/core/PlanNode.h"
+#include "velox/exec/fuzzer/ResultVerifier.h"
+#include "velox/exec/tests/utils/AssertQueryBuilder.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/tests/utils/VectorMaker.h"
+
+namespace facebook::velox::exec::test {
+
+// For aggregation calls min_by/max_by(x, y) and min_by/max_by(x, y, n),
+// MinMaxByResultVerifier works by computing an expected list of x-value-list
+// bucketed by the y values and checking the actual result values fall in the
+// expected list of buckets in order.
+// For example: for SELECT min_by(x, y, 2)
+//                  FROM (VALUES (2, 'b'), (1, 'b'), (3, 'a')) t(x, y),
+// the result verifier computes an expected list [[3], [2, 1]]. Suppose the
+// actual result is [3, 1], the verifier then checks that the result value 3
+// matches the first expected bucket [3] and the result value 1 falls in the
+// second expected bucket [2, 1]. The verifier reports an error if an actual
+// result value is not found in the first unexhausted expected bucket.
+//
+// min_by/max_by(x, y) and min_by/max_by(x, y, n) can return NULL in three
+// situations:
+// (1) all x in the group are masked out,
+// (2) all y in the group are NULL,
+// (3) for min_by/max_by(x, y), one of x associated with min/max y is NULL.
+class MinMaxByResultVerifier : public ResultVerifier {
+ public:
+  explicit MinMaxByResultVerifier(bool minBy) : minBy_{minBy} {}
+
+  bool supportsCompare() override {
+    return false;
+  }
+
+  bool supportsVerify() override {
+    return true;
+  }
+
+  void initialize(
```

Comment:
Would be nice move some of the large function into cpp file to reduce compilation time